File: Data-Structures-and-Algorithms/watercontainer/watercontainer.py
class Mutations:
    def maxArea(self, h):
        i, j = 0, len(h) - 1
        maxarea = 0
        while i < j:
            width = j-i
            maxarea = max(maxarea, min(h[i], h[j]) * width)
            if h[i] <= h[j]:
                i += 1
            else:
                j -= 1               
        return maxarea

height = [1, 8, 9, 5, 1200, 3, 1900, 300, 25]
print(height)

mutations = Mutations()
print(Mutations().maxArea(height))
print(mutations.maxArea(height))

# O(N) for time (N/2 steps to iterate through)
# O(1) for space (constant extra space, maxarea stores one information)

File: Data-Structures-and-Algorithms/watercontainer/watercontainer.js
var height = [1, 8, 9, 5, 1200, 3, 1900, 300, 25];
console.log(height);

const maxArea = function(h) {
    let i = 0;
    let j = h.length - 1;
    let maxarea = 0;
    while(i < j) {
        maxarea = Math.max((maxarea),(Math.min(h[i], h[j]) * (j-i)));
        h[i] <= h[j] ? i++ : j--
    }
    return maxarea
};

console.log(maxArea(height));


// O(N) for time (N/2 steps to iterate through)
// O(1) for space (constant extra space, maxarea stores one information)